File: frontend/src/lib/ai/mcts.ts
// PUCT Monte Carlo Tree Search guided by the AZNet — a port of backend/mcts.py.
// Dirichlet root noise is omitted (inference/exploitation only), matching
// ai.get_ai_move(..., add_dirichlet=False) on the backend.

import type { InferenceSession } from "onnxruntime-web";
import { State, Move, applyMove, getWinner, moveKey } from "./engine";
import { predict } from "./net";

const C_PUCT = 1.5;

class Node {
  prior: number;
  N = 0;
  W = 0;
  Q = 0;
  children = new Map<string, Node>();
  moves = new Map<string, Move>();
  state: State;
  terminalValue: number | null = null;

  constructor(prior: number, state: State) {
    this.prior = prior;
    this.state = state;
  }
}

/** Value from the perspective of the player about to move at `state`. */
function terminalValue(state: State): number | null {
  const w = getWinner(state);
  if (w === null) return null;
  if (w === "D") return 0;
  return state.player === w ? 1 : -1;
}

async function expand(node: Node, session: InferenceSession): Promise<number> {
  const tv = terminalValue(node.state);
  if (tv !== null) {
    node.terminalValue = tv;
    return tv;
  }
  const { moves, probs, value } = await predict(session, node.state);
  for (let i = 0; i < moves.length; i++) {
    const k = moveKey(moves[i]);
    node.children.set(k, new Node(probs[i], applyMove(node.state, moves[i])));
    node.moves.set(k, moves[i]);
  }
  return value;
}

function select(node: Node): Node {
  let bestScore = -Infinity;
  let best: Node | null = null;
  const sqrtSum = Math.sqrt(node.N + 1e-8);
  for (const child of node.children.values()) {
    const u = (C_PUCT * child.prior * sqrtSum) / (1 + child.N);
    // Child Q is from the opponent's perspective; flip for us.
    const score = -child.Q + u;
    if (score > bestScore) {
      bestScore = score;
      best = child;
    }
  }
  return best as Node;
}

/** Run MCTS and return visit counts keyed by "bi,ci", plus the move lookup. */
export async function runMcts(
  rootState: State,
  session: InferenceSession,
  simulations = 128,
): Promise<{ counts: Map<string, number>; moves: Map<string, Move> }> {
  const root = new Node(1, rootState);
  await expand(root, session);

  for (let s = 0; s < simulations; s++) {
    let node = root;
    const path: Node[] = [node];

    while (node.children.size > 0 && node.terminalValue === null) {
      node = select(node);
      path.push(node);
    }

    const value = node.terminalValue !== null ? node.terminalValue : await expand(node, session);

    // Backup: value is from the perspective of the player to move at `node.state`;
    // the perspective flips at each step up the tree.
    let v = value;
    for (let i = path.length - 1; i >= 0; i--) {
      const n = path[i];
      n.N += 1;
      n.W += v;
      n.Q = n.W / n.N;
      v = -v;
    }
  }

  const counts = new Map<string, number>();
  for (const [k, child] of root.children) counts.set(k, child.N);
  return { counts, moves: root.moves };
}

File: backend/mcts.py
"""PUCT Monte Carlo Tree Search guided by the AZNet."""
import math
import numpy as np

from engine import legal_moves, apply_move, winner, state_key
from net import predict

C_PUCT = 1.5


class Node:
    __slots__ = ("prior", "N", "W", "Q", "children", "state", "terminal_value")

    def __init__(self, prior, state):
        self.prior = prior          # P(s,a) from parent's perspective
        self.N = 0
        self.W = 0.0
        self.Q = 0.0
        self.children = {}          # move -> Node
        self.state = state          # state AFTER the move that led here
        self.terminal_value = None  # value from the perspective of player-to-move at `state`


def _terminal_value(state):
    """Return value from the perspective of the player about to move at `state`.
    If state has a winner, the previous player just won, so the player to move loses.
    """
    w = winner(state)
    if w is None:
        return None
    if w == "D":
        return 0.0
    # `state[3]` is the player to move; if they don't match the winner, they lost.
    return 1.0 if state[3] == w else -1.0


def _expand(node, net, device):
    tv = _terminal_value(node.state)
    if tv is not None:
        node.terminal_value = tv
        return tv
    probs, value = predict(net, node.state, device=device)
    for move, p in probs.items():
        child_state = apply_move(node.state, move)
        node.children[move] = Node(prior=p, state=child_state)
    return value


def _select(node):
    best_score, best_move, best_child = -float("inf"), None, None
    sqrt_sum = math.sqrt(node.N + 1e-8)
    for move, child in node.children.items():
        u = C_PUCT * child.prior * sqrt_sum / (1 + child.N)
        # Child Q is from opponent's perspective; flip for us.
        score = -child.Q + u
        if score > best_score:
            best_score, best_move, best_child = score, move, child
    return best_move, best_child


def run_mcts(root_state, net, simulations=64, device="cpu", add_dirichlet=True):
    root = Node(prior=1.0, state=root_state)
    _expand(root, net, device)

    # Dirichlet noise on root priors for exploration during self-play.
    if add_dirichlet and root.children:
        moves = list(root.children.keys())
        noise = np.random.dirichlet([0.3] * len(moves))
        for m, n in zip(moves, noise):
            root.children[m].prior = 0.75 * root.children[m].prior + 0.25 * n

    for _ in range(simulations):
        node = root
        path = [node]
        # Selection: walk to a leaf
        while node.children and node.terminal_value is None:
            _, node = _select(node)
            path.append(node)

        # Expansion / evaluation
        if node.terminal_value is not None:
            value = node.terminal_value
        else:
            value = _expand(node, net, device)

        # Backup. `value` is from the perspective of the player to move at `node.state`.
        # As we walk up, the perspective flips each step.
        v = value
        for n in reversed(path):
            n.N += 1
            n.W += v
            n.Q = n.W / n.N
            v = -v

    # Visit-count policy
    counts = {m: c.N for m, c in root.children.items()}
    return counts


def policy_from_counts(counts, temperature=1.0):
    moves = list(counts.keys())
    visits = np.array([counts[m] for m in moves], dtype=np.float64)
    if temperature <= 1e-3:
        probs = np.zeros_like(visits)
        probs[visits.argmax()] = 1.0
    else:
        v = visits ** (1.0 / temperature)
        probs = v / v.sum()
    return moves, probs